### PR TITLE
Added infix expressions.

### PR DESCRIPTION
**Pull request by AndrewMcBurney**
#3, #26

### Changes
- [x] +
- [x] -
- [x] *
- [x] /
- [x] %

### Usage

Adds `+`, `-`, `*`, `/`, and `%` operators:

```js
// Declare references
let lhs = new cgl.Reference(
  new cgl.InterfaceVariable(
    cgl.Qualifier.In, new cgl.VariableSource(new cgl.Type(cgl.Kind.Int), '1')
  )
);
let rhs = new cgl.Reference(
  new cgl.InterfaceVariable(
    cgl.Qualifier.In, new cgl.VariableSource(new cgl.Type(cgl.Kind.Float), '1.5')
  )
);
let rhsInt = new cgl.Reference(
  new cgl.InterfaceVariable(
    cgl.Qualifier.In, new cgl.VariableSource(new cgl.Type(cgl.Kind.Int), '1')
  )
);

// Addition expression
// 1 + 1.5
new cgl.Addition(lhs, rhs);

// Subtraction expression
// 1 - 1.5
new cgl.Subtraction(lhs, rhs);

// Multiplication expression
// 1 * 1.5
new cgl.Multiplication(lhs. rhs);

// Division expression
// 1 / 1.5
new cgl.Division(lhs, rhs);

// Modulo expression
// 1 % 1
new cgl.Modulo(lhs, rhsInt);
```

### CODE REVIEW

**Comment by AndrewMcBurney** [comment]
I'm going to fix up this branch to account for @abhimadan's most recent changes first.

File: src/expressions/math/infix/infix_expression.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import Set from '../../../util/set';
+import Type from '../../../type';
+import Kind from '../../../kind';
+
+export default abstract class InfixExpression implements Expression {
+    protected lhs: Expression;
+    protected rhs: Expression;
+
+    constructor(lhs: Expression, rhs: Expression) {
+        if (!(lhs.returnType().checkEquals(new Type(Kind.Int)) || lhs.returnType().checkEquals(new Type(Kind.Float))) ||
```
Comment (abhimadan) [inline src/expressions/math/infix/infix_expression.ts]: What about vector operations? At least addition and subtraction between vectors and matrices of the same size are possible (though multiplication won't be possible until we improve the type system to handle vectors and matrices better).

Comment (abhimadan) [reply src/expressions/math/infix/infix_expression.ts]: Also, modulo only works on two ints, so maybe we want to check that at the subclass level?

Comment (AndrewMcBurney) [reply src/expressions/math/infix/infix_expression.ts]: Good points! I'll make both of those changes.

File: src/expressions/math/infix/infix_expression.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+export default abstract class InfixExpression implements Expression {
+    protected lhs: Expression;
+    protected rhs: Expression;
+
+    constructor(lhs: Expression, rhs: Expression) {
+        this.lhs = lhs;
+        this.rhs = rhs;
+
+        if (!(this.bothSameSizeVectors() || this.bothSameSizeMatrices()) && !this.bothSidesScalarTypes()) {
```
Comment (abhimadan) [inline src/expressions/math/infix/infix_expression.ts]: nit: can you change it to read `if (!cond1 && !cond2 && !cond3)`? I'm finding the negation of the OR condition to be difficult to read.

File: src/expressions/math/infix/multiplication.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import InfixExpression from './infix_expression'
+import Expression from '../../expression';
+
+export default class Multiplication extends InfixExpression {
+    constructor(lhs: Expression, rhs: Expression) {
```
Comment (abhimadan) [inline src/expressions/math/infix/multiplication.ts]: We still need to check here that the two expressions are the same scalar type.

File: src/expressions/math/infix/modulo.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import Expression from '../../expression';
+import InfixExpression from './infix_expression'
+import Kind from '../../../kind';
+import Type from '../../../type';
+
+export default class Modulo extends InfixExpression {
+    constructor(lhs: Expression, rhs: Expression) {
```
Comment (abhimadan) [inline src/expressions/math/infix/modulo.ts]: See comment on `multiplication.ts`. Also add back the check that the two expressions must both be ints.

File: src/expressions/math/infix/division.ts [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import InfixExpression from './infix_expression'
+import Expression from '../../expression';
+
+export default class Division extends InfixExpression {
+    constructor(lhs: Expression, rhs: Expression) {
```
Comment (abhimadan) [inline src/expressions/math/infix/division.ts]: See comment on `multiplication.ts`.

File: src/expressions/math/infix/multiplication.ts (hunk 3 above)
Comment (AndrewMcBurney) [reply src/expressions/math/infix/multiplication.ts]: Can't you multiply a float by an int and vice versa, or do they need to be explicitly casted?

Comment (abhimadan) [reply src/expressions/math/infix/multiplication.ts]: You can - my mistake. My main point was that we shouldn't allow matrices and vectors here.

Comment (AndrewMcBurney) [reply src/expressions/math/infix/multiplication.ts]: Ahh, good call

**Comment by abhimadan** [comment]
Now that Paul's PR is merged, can you change all `TypeError`s in this PR to `TypeException`s?

File: src/expressions/math/infix/division.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import InfixExpression from './infix_expression'
+import Expression from '../../expression';
+import TypeException from '../../../exceptions/typeexception';
+
+export default class Division extends InfixExpression {
+    constructor(lhs: Expression, rhs: Expression) {
+        super(lhs, rhs);
+
+        if (super.anySideVectorOrMatrix(lhs, rhs)) {
```
Comment (pbardea) [inline src/expressions/math/infix/division.ts]: Minor nit: name confused me for a sec. Something that would make more sense to me would be `eitherVectorOrMatrix`, but this probably just comes down to personal preference. Whichever ya prefer.

Comment (AndrewMcBurney) [reply src/expressions/math/infix/division.ts]: Sounds good to me! 👍

**Pull request merged**